api/python/t4/snapshots.py

```diff
@@ -131,10 +139,350 @@ def download_file_from_snapshot(src, dst, snapshothash):
     snapshot_data = read_snapshot_by_hash(bucket, snapshothash)
     obj_rec = snapshot_data['contents'][key]
     return download_file(src, dst, version=obj_rec['VersionId'])
-    
+
 
 def download_bytes_from_snapshot(src, snapshothash):
     bucket, key = split_path(src)
     snapshot_data = read_snapshot_by_hash(bucket, snapshothash)
     obj_rec = snapshot_data['contents'][key]
     return download_bytes(src, version=obj_rec['VersionId'])
+
+class PhysicalKeyType(Enum):
+    LOCAL = 1
+    S3 = 2
+
+def hash_file(path):
+    """ Returns SHA256 hash of file at path. """
+    with open(path, 'rb') as f:
+        buf = f.read(4096)
+        hasher = hashlib.sha256()
+        while buf:
+            hasher.update(buf)
+            buf = f.read(4096)
+
+        return hasher.hexdigest()
+
+def dereference_physical_key(physical_key):
+    ty = physical_key['type']
+    if ty == PhysicalKeyType.LOCAL.name:
+        return open(physical_key['path'])
+
+    raise NotImplementedError
+
+def get_package_registry(path):
+    """ Returns the package registry for a given path """
+    if path.startswith('s3://'):
+        bucket = path[5:].partition('/')[0]
+        return "s3://{}/.quilt/packages/".format(bucket)
+    else:
+        # local path
+        return "appdirs/.quilt/packages" # TODO: replace with actual logic
+
+class PackageEntry(object):
+    """
+    Represents an entry at a logical key inside a package.
+    """
+    __slots__ = ['physical_keys', 'size', 'hash_obj', 'meta']
+    def __init__(self, physical_keys, size, hash_obj, meta):
+        """
+        Creates an entry.
+
+        Args:
+            physical_keys([physical_key]): nonempty list of physical keys
+            size(number): size of object in bytes
+            hash({'type': string, 'value': string}): hash object
+            meta(dict): metadata dictionary
+
+        Returns:
+            a PackageEntry
+        """
+        self.physical_keys = physical_keys
+        self.size = size
+        self.hash = hash_obj
```

Comment:
`self.hash_obj = hash_obj`

I'd expect the current code to fail somewhere? Is it not being used?

Reply:
should be `self.hash`, fixed